### PR TITLE
Disabled automatic bidirectional one-to-many relationship generation when using neo4j

### LINKED ISSUE #11691: Neo4J: Cyclic OneToMany adds an incorrect ManyToOne
##### **Overview of the issue**

Neo4J generation
Cyclic Relationship to entity itself results in second, incorrect property.

##### **Motivation for or Use Case**

I want a Person to have friends as a collection of type Person.

##### **Reproduce the error**

    entity Person {
       name: String
    }

    relationship OneToMany {
       Person{friends} to Person
    }

Generate application based on this. Generated domain class:

    @Node
    public class Person implements Serializable {

        // correct: 

        @Relationship
        private Set<Person> friends = new HashSet<>();

        // incorrect:

        @Relationship("person")
        @JsonIgnoreProperties("friends")
         private Person person;

    }

Also, some additional code is generated based on this, including CRUD in the front end.

##### **Related issues**

This similar issue was related to React, and Neo4J was not used back then:
https://github.com/jhipster/generator-jhipster/issues/11549

##### **Suggest a Fix**

Do not generate the reverse relationship in a OneToMany to type itself for Neo4J.

##### **JHipster Version(s)**

6.8.0

##### **JHipster configuration**

{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.blabla",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "6.8.0",
    "applicationType": "monolith",
    "baseName": "blabla",
    "packageName": "com.blabla",
    "packageFolder": "com/blabla",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "enableHibernateCache": false,
    "websocket": false,
    "databaseType": "neo4j",
    "devDatabaseType": "neo4j",
    "prodDatabaseType": "neo4j",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "",
    "embeddableLaunchScript": false,
    "useSass": true,
    "clientPackageManager": "npm",
    "clientFramework": "angularX",
    "clientTheme": "minty",
    "clientThemeVariant": "primary",
    "creationTimestamp": 1588343391509,
    "testFrameworks": ["cucumber"],
    "jhiPrefix": "jhi",
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": ["en"],
    "blueprints": [],
    "herokuAppName": "blabla",
    "herokuDeployType": "jar"
  }
}
---
If you have a JDL please wrap it in below structure
  <details>
  <summary>JDL definitions</summary>
  <pre>
     JDL content here
  </pre>
  </details>
-->

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

Here, Person is called Trainer. I think I already see the problem. There is a ManyToOne after the OneToMany that should probably go. I checked, and it is not in the jhipster-jdl.jh

{
    "name": "Trainer",
    "fields": [
        {
            "fieldName": "name",
            "fieldType": "String"
        },
        {
            "fieldName": "code",
            "fieldType": "Long"
        },
        {
            "fieldName": "level",
            "fieldType": "Integer"
        },
        {
            "fieldName": "xp",
            "fieldType": "Integer"
        },
        {
            "fieldName": "team",
            "fieldType": "Team",
            "fieldValues": "VALOR,MYSTIC,INSTINCT"
        }
    ],
    "relationships": [
        {
            "relationshipType": "one-to-many",
            "otherEntityName": "trainer",
            "otherEntityRelationshipName": "trainer",
            "relationshipName": "friends"
        },
        {
            "relationshipType": "many-to-one",
            "otherEntityName": "trainer",
            "otherEntityRelationshipName": "friends",
            "relationshipName": "trainer",
            "otherEntityField": "id"
        }
    ],
    "changelogDate": "20200502070200",
    "entityTableName": "trainer",
    "dto": "no",
    "pagination": "no",
    "service": "serviceImpl",
    "jpaMetamodelFiltering": false,
    "fluentMethods": true,
    "readOnly": false,
    "embedded": false,
    "clientRootFolder": "",
    "applications": "*"
}

##### **Browsers and Operating System**

Linux Mint 19

-   [X] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by MathieuAA**
Fixes #11691 

---

Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Review by atomfrede — APPROVED** [review]

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs [hunk 1]
```diff
@@ -517,8 +517,10 @@ public class <%= asEntity(entityClass) %> implements Serializable {
             <%_ if (databaseType === 'couchbase' && !otherEntityIsEmbedded) { _%>
         this.<%= relationshipFieldName %>Ids.add(<%= otherEntityName %>.getId());
             <%_ } _%>
-            <%_ if (relationshipType === 'one-to-many') { _%>
-        <%= otherEntityName %>.set<%= otherEntityRelationshipNameCapitalized %>(this);
+            <%_ if (relationshipType === 'one-to-many') {
+                if (databaseType !== 'neo4j') { _%>
+                    <%= otherEntityName %>.set<%= otherEntityRelationshipNameCapitalized %>(this);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs:522]:
```suggestion
        <%= otherEntityName %>.set<%= otherEntityRelationshipNameCapitalized %>(this);
```

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs [hunk 2]
```diff
@@ -533,8 +535,10 @@ public class <%= asEntity(entityClass) %> implements Serializable {
             <%_ if (databaseType === 'couchbase' && !otherEntityIsEmbedded) { _%>
         this.<%= relationshipFieldName %>Ids.remove(<%= otherEntityName %>.getId());
             <%_ } _%>
-            <%_ if (relationshipType === 'one-to-many') { _%>
-        <%= otherEntityName %>.set<%= otherEntityRelationshipNameCapitalized %>(null);
+            <%_ if (relationshipType === 'one-to-many') {
+                if (databaseType !== 'neo4j') { _%>
+                    <%= otherEntityName %>.set<%= otherEntityRelationshipNameCapitalized %>(null);
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs:540]:
```suggestion
        <%= otherEntityName %>.set<%= otherEntityRelationshipNameCapitalized %>(null);
```

**Review by mshima — APPROVED** [review]
Indentation should follow java sources instead of ejs source.

**Comment by MathieuAA** [comment]
Thanks for the review. I'll fix the remaining things in later in the afternoon

File: test/jdl/jdl-importer.spec.js [hunk 3]
```diff
@@ -1937,5 +1937,30 @@ paginate * with infinite-scroll
...
+            let importState;
+
+            before(() => {
+                const content = `entity Person {
+   name String
+}
+
+relationship OneToMany {
+   Person{friends} to Person
+}`;
```
Comment (qmonmert) [inline test/jdl/jdl-importer.spec.js:1950]: indentation 😇

Comment (MathieuAA) [reply test/jdl/jdl-importer.spec.js:1950]: There's no issue :)

Comment (qmonmert) [reply test/jdl/jdl-importer.spec.js:1950]: :o

Comment (MathieuAA) [reply test/jdl/jdl-importer.spec.js:1950]: Is it the string templating thing that's misleading? If so, I'll extract the JDL content to a file.

**Pull request merged**